src/Development/IDE/Core/FileExists.hs

```diff
@@ -0,0 +1,187 @@
+{-# LANGUAGE OverloadedLists      #-}
+{-# LANGUAGE TypeFamilies         #-}
+{-# LANGUAGE UndecidableInstances #-}
+module Development.IDE.Core.FileExists
+  ( fileExistsRules
+  , modifyFileExists
+  , getFileExists
+  )
+where
+
+import           Control.Concurrent.Extra
+import           Control.Exception
+import           Control.Monad.Extra
+import qualified Data.Aeson                    as A
+import           Data.Binary
+import qualified Data.ByteString               as BS
+import           Data.Map.Strict                ( Map )
+import qualified Data.Map.Strict               as Map
+import           Data.Maybe
+import qualified Data.Text                     as T
+import           Development.IDE.Core.FileStore
+import           Development.IDE.Core.Shake
+import           Development.IDE.Types.Location
+import           Development.Shake
+import           Development.Shake.Classes
+import           GHC.Generics
+import           Language.Haskell.LSP.Messages
+import           Language.Haskell.LSP.Types
+import           Language.Haskell.LSP.Types.Capabilities
+import qualified System.Directory as Dir
+
+-- | A map for tracking the file existence
+type FileExistsMap = (Map NormalizedFilePath Bool)
+
+-- | A wrapper around a mutable 'FileExistsMap'
+newtype FileExistsMapVar = FileExistsMapVar (Var FileExistsMap)
+
+instance IsIdeGlobal FileExistsMapVar
+
+-- | Grab the current global value of 'FileExistsMap' without acquiring a dependency
+getFileExistsMapUntracked :: Action FileExistsMap
+getFileExistsMapUntracked = do
+  FileExistsMapVar v <- getIdeGlobalAction
+  liftIO $ readVar v
+
+-- | Modify the global store of file exists
+modifyFileExistsAction :: (FileExistsMap -> IO FileExistsMap) -> Action ()
+modifyFileExistsAction f = do
+  FileExistsMapVar var <- getIdeGlobalAction
+  liftIO $ modifyVar_ var f
+
+-- | Modify the global store of file exists
+modifyFileExists :: IdeState -> [(NormalizedFilePath, Bool)] -> IO ()
+modifyFileExists state changes = do
+  FileExistsMapVar var <- getIdeGlobalState state
+  changesMap           <- evaluate $ Map.fromList changes
+
+  -- Mask to ensure that the previous values are flushed together with the map update
+  mask $ \_ -> do
+    modifyVar_ var $ evaluate . Map.union changesMap
+    let flushPreviousValues = do
+          liftIO $ mapM_ (deleteValue state GetFileExists . fst) changes
+    void $ shakeRun state [flushPreviousValues]
```

Comment:
```suggestion
    flushPreviousValues
```
You don’t need to call `shakeRun` here. You can just call `flushPreviousValues` which avoids potential issues from masking `shakeRun` (not entirely sure if that causes dissues but I’d rather avoid it if we can).

Reply:
I thought I had done this. Thanks!